File: src/authoring/schema.py
"""The declarative scenario-file schema (pydantic, the config boundary).

Validates the shape of an authored scenario file into typed specs the interpreter
lowers to frozen engine objects. Step 0 = the **composition subset**: stocks (ICs),
flows (existing frozen types + wiring + a param-set reference), constant forcings,
and the run config (integrator/dt/steps). Every model is ``extra="forbid"`` — a
typo'd key is a schema error, not a silent drop (the ``config`` loader discipline).

**The pyyaml numeric hazard is handled here.** ``yaml.safe_load`` uses the YAML-1.1
float resolver, under which ``1e-3`` (no decimal point) parses as a *string*, not a
float. Every float field below is a pydantic ``float`` (string→float coercion is a
backstop), and scenario files are written with an explicit decimal point
(``1.0e-3``); the interpreter's anchor test additionally asserts the loaded ICs and
forcings equal the frozen reference values, so a silent string-parse fails loudly.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field, model_validator


class StockSpec(BaseModel):
    """One stock's declaration: id, namespace, quantity, kind, initial amount.

    ``quantity`` is a :class:`simcore.quantities.Quantity` value ("carbon",
    "water", "nitrogen", "oxygen", "energy"); ``kind`` a
    :class:`simcore.quantities.StockKind` value ("pool", "population",
    "boundary"). ``composition`` (optional) maps quantity values → coeffs for a
    multi-quantity stock (e.g. CO2 = {"carbon": 1, "oxygen": 2}); omitted means the
    frozen 1:1 default. ``unclamped`` / ``extinction_threshold`` mirror the frozen
    ``Stock`` fields (meaningful only for BOUNDARY sources / POPULATION stocks).
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    domain: str
    quantity: str
    kind: str
    amount: float | str
    """The initial amount: a literal float, or a **template expression** (Step 3).

    A string is a bounded-grammar expression over the scenario's declared
    ``parameters`` (e.g. ``"param('crew_count') * 1000.0"``), evaluated to a literal at
    interpret time (:func:`authoring.template.eval_numeric_field`). A bare float is the
    Step-0 form, unchanged — all-literal scenarios still lower identically.
    """
    composition: dict[str, float] | None = None
    unclamped: bool = False
    extinction_threshold: float = 0.0


class ParamPackRef(BaseModel):
    """A reference to a **parameter pack** — an alternate param file (Step 1).

    ``pack`` is a path to a param file in the frozen loader's own
    ``{value, unit, source}`` schema, resolved **relative to the scenario file's
    directory** (so a modder-shipped scenario+pack bundle relocates intact). The
    *same* frozen loader reads it, so a pack's values pass the frozen schema/bounds/
    unit validation — a pack is a param file, not a way around the guards. The
    "cultivar = param pack" species primitive: an author overrides a value (e.g. a
    metabolic fraction) by shipping a full pack file with the changed value.
    """

    model_config = ConfigDict(extra="forbid")

    pack: str


class KineticsSpec(BaseModel):
    """An **authored kinetics** flow: a rate expression × a stoichiometry (Step 2).

    The declarative alternative to selecting a frozen flow ``type``. ``rate`` is a
    bounded-grammar expression string (see
    :func:`authoring.expr_parser.parse_rate_expr`) giving the **instantaneous** rate —
    ``dt``-independent (the grammar has no ``dt``
    token, so RK4-order-safety is structural). ``stoichiometry`` maps each touched
    stock id → its (integer/rational) coefficient; the flow emits ``coeff · rate · dt``
    per leg, so it is **balanced by construction** for any rate value provided the
    coefficient vector balances per quantity — which the interpreter verifies against
    the stock compositions at build time (decision C).

    **Authored ≠ validated** (decision B): conservation + determinism are guaranteed,
    scientific validity is the author's responsibility — an authored-kinetics run
    carries no calibration claim, no golden, no manifest entry.
    """

    model_config = ConfigDict(extra="forbid")

    rate: str
    stoichiometry: dict[str, float]


class FlowSpec(BaseModel):
    """One flow: either a frozen flow *type* + wiring, or authored *kinetics*.

    Exactly one of ``type`` (a key into :data:`authoring.flow_registry.FLOW_TYPES`, the
    author-selectable frozen-flow surface — with ``wiring`` mapping the constructor's
    stock-id fields → declared stock ids) **xor** ``kinetics`` (an authored
    rate×stoichiometry flow, Step 2) must be given; the model validator enforces the
    exclusivity and that a kinetics flow carries no ``wiring`` (its stoichiometry names
    stocks directly). ``params`` selects the params object: a **string** names a
    frozen param set (a frozen flow type's default set — the Step-0 form; or, for a
    kinetics flow, a :data:`authoring.flow_registry.PARAM_LOADERS` key whose loaded
    values the rate's ``param("…")`` reads, so authored params still pass the frozen
    bounds/unit guards), or a :class:`ParamPackRef` supplies a parameter pack (Step 1).
    It must be omitted for a param-free flow. Partial-merge/bundling packs are deferred.
    """

    model_config = ConfigDict(extra="forbid")

    id: str
    type: str | None = None
    priority: int = 0
    wiring: dict[str, str] = Field(default_factory=dict)
    kinetics: KineticsSpec | None = None
    params: str | ParamPackRef | None = None
    rate_class: str = "fast"
    """The **rate class**: which side of the multi-rate partition this flow steps on.

    ``"fast"`` (the default) sub-steps ``n_sub`` times at ``dt/n_sub`` *inside* each
    master step; ``"slow"`` steps once at the master ``dt`` (as two ``dt/2`` halves
    under Strang splitting). The legal values are validated by the interpreter, not here
    (:data:`authoring.interpreter._RATE_CLASSES`) — the ``integrator``/``quantity``/
    ``kind`` discipline, where an unknown value is an ``AuthoringError`` rather than a
    pydantic error.

    **Spelled ``rate_class``, not ``rate``, deliberately.** ``KineticsSpec.rate`` — one
    nesting level down, on the same flow — is the *rate law* (how fast the flow runs);
    this is the *rate class* (which cadence it is integrated on). A flow may carry both,
    so a bare ``rate:`` here would put two senses of the word in one block. The name
    also matches what the concept is called everywhere else it appears:
    :data:`authoring.interpreter._RATE_CLASSES` and the manifest's ``rate_classes``.

    **This is a *property*, not a *reference* — which is exactly why it lives on the
    flow.** The rate class travels *with* the flow, inside a bundle, and ``compose``'s
    ``{bundle, prefix}`` rewrite cannot touch it, because prefixing rewrites **ids**.
    The rejected alternative — a top-level ``fast: [flow-id, …]`` list — would have been
    a list of id *references*, i.e. a new rewrite surface in ``compose.apply_includes``
    that silently mis-fires the moment someone forgets it. This shape has zero
    referential-integrity surface.

    **Defaulting to ``"fast"`` is load-bearing, not a convenience.** ``rate_class``
    defaulting to fast + ``n_sub`` defaulting to 1 ⇒ an **empty slow set** ⇒ a file with
    no multi-rate keys lowers to today's single-rate trajectory *by construction* (the
    identity path, measured in ``tests/test_authoring_multirate_identity.py``).
    Defaulting to ``"slow"`` could not do this: with an empty fast set, Strang would run
    two ``dt/2`` half-steps, which is **not** one full Euler step.
    """

    @model_validator(mode="after")
    def _check_type_xor_kinetics(self) -> FlowSpec:
        if (self.type is None) == (self.kinetics is None):
            raise ValueError(
                f"flow {self.id!r}: give exactly one of 'type' (a frozen flow type) "
                f"or 'kinetics' (an authored rate×stoichiometry flow)"
            )
        if self.kinetics is not None and self.wiring:
            raise ValueError(
                f"flow {self.id!r}: an authored 'kinetics' flow takes no 'wiring' "
                f"(its stoichiometry names stocks directly)"
            )
        return self


class ForcingSpec(BaseModel):
    """A forcing schedule. Step 0 = constant forcings only (``const``).

    Computed schedules (the Power half-sine, biosphere weather) are a later step;
    the Step-0 composition anchor (crew) uses only constant forced rates. ``const``
    is a literal float or a **template expression** over the declared ``parameters``
    (Step 3), evaluated to a literal at interpret time (like a stock ``amount``).
    """

    model_config = ConfigDict(extra="forbid")

    const: float | str


class IncludeSpec(BaseModel):
    """A **prefixed** bundle include — a bundle instance under a namespace (Step 6c).

    The bare-string include form (``- bundles/battery.domain.yaml``) merges a bundle's
    ids verbatim; a mapping form (``{bundle: <path>, prefix: <name>}``) **namespaces**
    every id the bundle declares — each stock id, flow id and forcing key becomes
    ``<prefix>.<id>``, and every reference to them (``wiring`` values, ``stoichiometry``
    keys, and the ``stock(...)``/``forcing(...)`` refs inside a ``kinetics`` rate) is
    rewritten to match (:func:`authoring.compose.apply_includes`). This is what lets the
    **same** bundle be included more than once without the id collision a bare double
    include hits (:func:`authoring.compose` documents the surface).

    **Only disjoint / kinetics-style bundles are multi-instanceable.** A frozen flow
    that binds a forcing by a *hardcoded* name (the crew flows read ``crew_o2_intake``
    etc. from a frozen module constant, not through wiring) cannot find a namespaced
    forcing key — a prefixed crew include would raise a resolve-time ``KeyError``.
    Prefixing a forcing-bound frozen bundle is therefore an unsupported (documented)
    case, the crew analogue of the greenhouse's hardcoded ``CARBON_POOL`` blocker.
    Bundle **parameter** namespacing is likewise deferred (the only param-bearing
    bundle, crew, is blocked for the forcing reason anyway) — two prefixed instances of
    a param-bearing bundle collide on the parameter name, the honest boundary.
    """

    model_config = ConfigDict(extra="forbid")

    bundle: str
    prefix: str = Field(min_length=1)


class BundleSpec(BaseModel):
    """A reusable **domain / species bundle**: stocks + flows + forcings (Step 6).

    A *domain* is a named stock+flow bundle over a quantity set; a *species* is a
    flow-set + param-set + stock-template (a bundle whose stock ``amount``s are
    template expressions over its own ``parameters`` — e.g. the crew "species"
    parametrized by ``crew_count``). A scenario **includes** one or more bundle files
    (:attr:`ScenarioSpec.includes`); the interpreter merges each bundle's parameters/
    stocks/flows/forcings into the scenario's flat graph
    (:func:`authoring.compose.apply_includes`).

    A bundle carries **no run config** (integrator/dt/steps/name/rng_seed) and **no
    nested** ``includes`` — both are rejected by ``extra="forbid"`` (run config lives
    only in the top-level scenario; includes are flat, one level deep — Step 6 scope).
    """

    model_config = ConfigDict(extra="forbid")

    parameters: dict[str, float] = Field(default_factory=dict)
    stocks: list[StockSpec] = Field(default_factory=list)
    flows: list[FlowSpec] = Field(default_factory=list)
    forcings: dict[str, ForcingSpec] = Field(default_factory=dict)


class ScenarioSpec(BaseModel):
    """A whole authored scenario: run config + parameters + stocks + flows + forcings.

    ``integrator`` is "euler" or "rk4"; ``dt`` the step size (seconds); ``steps``
    the run length; ``rng_seed`` the state seed (default 0). The interpreter builds a
    no-reset graph from this (annual-reset/phenology remains out of scope).

    ``dt`` + ``n_sub`` + each flow's ``rate_class`` are the **multi-rate** knob
    (post-roadmap): ``dt`` is the **coupling cadence** at which a step commits and peers
    read; the fast flow set sub-steps internally at ``dt/n_sub``. Omitting both keys is
    the single-rate path — an empty slow set at ``n_sub = 1`` — which reproduces the
    pre-multi-rate trajectory bit-for-bit. The one refused combination is ``n_sub = 1``
    *with* a non-empty slow set: it buys no rate separation and no performance win, yet
    silently moves the answer (see :func:`authoring.interpreter.interpret`).

    ``parameters`` (Step 3) is the **template contract**: named scalars with default
    values that an instantiation may override (``interpret(..., overrides=…)``), and
    that a stock ``amount`` / forcing ``const`` expression may read via ``param('…')``.
    Empty (the Step-0 form) means a fully-literal scenario with no knobs.

    ``includes`` (Step 6) is a list of bundle includes, resolved relative to the
    scenario file's directory. Each element is either a bare **bundle-file path** (a
    verbatim merge — each a :class:`BundleSpec`) or a :class:`IncludeSpec`
    (``{bundle, prefix}`` — a **namespaced** instance, Step 6c). Each included bundle's
    parameters/stocks/flows/forcings merge into this scenario's
    (:func:`authoring.compose.apply_includes`, run at the top of ``interpret``): a
    scenario is thus *composed* from reusable domain/species bundles + its own inline
    declarations. A duplicate id/key/parameter across any two sources is an
    ``AuthoringError`` (no silent override). Empty (the pre-Step-6 form) means a
    self-contained scenario.
    """

    model_config = ConfigDict(extra="forbid")

    name: str
    integrator: str
    dt: float
    steps: int
    n_sub: int = Field(default=1, ge=1)
    """How many sub-steps the **fast** flow set takes inside one master ``dt`` step.

    This is the knob that turns ``dt`` into a **coupling cadence** — what a neighbouring
    domain sees exported — rather than what the fast flows solve at, which becomes the
    **effective sub-step** ``dt/n_sub``. Default 1 (with the default all-fast partition)
    is the single-rate path, bit-identical to the pre-multi-rate engine.

    ``ge=1`` mirrors ``simcore.multirate.multirate_step``'s own ``n_sub >= 1`` guard at
    the file boundary, so a nonsensical value is a schema error on the author's file
    rather than a ``ValueError`` surfacing from the core mid-run.

    **``n_sub`` does not by itself make a step safe** — it only makes ``dt/n_sub`` the
    number that must satisfy ``k·dt < 1``. Measured: ``n_sub=2`` at ``dt=3600`` still
    wrecks the cabin (36.0 against a truth of 8.0). Multi-rate is the *performance
    enabler*, not the hazard closer; the build-time precondition on the **effective**
    sub-step is a later step.
    """
    rng_seed: int = 0
    includes: list[str | IncludeSpec] = Field(default_factory=list)
    parameters: dict[str, float] = Field(default_factory=dict)
    stocks: list[StockSpec] = Field(default_factory=list)
    flows: list[FlowSpec] = Field(default_factory=list)
    forcings: dict[str, ForcingSpec] = Field(default_factory=dict)

File: rust/crates/authoring/src/schema.rs
//! The declarative scenario-file schema — the Rust mirror of Python
//! `authoring.schema` (Phase 9, Step 4b).
//!
//! Binds a [`crate::yaml::YamlValue`] tree (from the closed-subset reader) into typed
//! specs the interpreter lowers to frozen engine objects, exactly as the pydantic
//! models do on the Python side. Every mapping is **`extra="forbid"`** — an unknown
//! key is a schema error, not a silent drop — and the `type`-xor-`kinetics` flow
//! exclusivity is enforced here (mirroring `FlowSpec._check_type_xor_kinetics`).
//!
//! **The pyyaml numeric hazard is handled in [`crate::yaml::is_yaml_number`].** A
//! stock `amount` / forcing `const` is a `number | expression` union: a **bare**
//! YAML-1.1 number lowers to [`NumericField::Number`]; a quoted scalar or a non-number
//! bare scalar is a template [`NumericField::Expr`] — the same split pyyaml makes when
//! it returns a `float` vs a `str`.

use crate::errors::AuthoringError;
use crate::yaml::{is_yaml_number, YamlValue};

/// A numeric scenario field (`StockSpec.amount`, `ForcingSpec.const`): a literal
/// number, or a template expression over the scenario's `parameters` (Step 3).
#[derive(Debug, Clone, PartialEq)]
pub enum NumericField {
    /// A bare YAML-1.1 number, already parsed to `f64` (correctly-rounded via
    /// `f64::from_str`, matching Python `float()`).
    Number(f64),
    /// A template expression string (a quoted scalar or a non-number bare scalar),
    /// parsed + evaluated to a literal at interpret time (Step 3 boundary-eval).
    Expr(String),
}

/// One stock's declaration (mirrors `StockSpec`).
#[derive(Debug, Clone, PartialEq)]
pub struct StockSpec {
    pub id: String,
    pub domain: String,
    pub quantity: String,
    pub kind: String,
    pub amount: NumericField,
    pub composition: Option<Vec<(String, f64)>>,
    pub unclamped: bool,
    pub extinction_threshold: f64,
}

/// An authored-kinetics flow's rate × stoichiometry (mirrors `KineticsSpec`).
#[derive(Debug, Clone, PartialEq)]
pub struct KineticsSpec {
    pub rate: String,
    pub stoichiometry: Vec<(String, f64)>,
}

/// A flow's params selector (mirrors the `str | ParamPackRef | None` union).
#[derive(Debug, Clone, PartialEq)]
pub enum ParamsSpec {
    /// A frozen param-set name (`params: crew`).
    Named(String),
    /// A parameter pack (`params: {pack: …}`) — **deferred in the Rust port** (Step 4b
    /// keeps named sets only; a pack surfaces as an interpret-time "deferred" error).
    Pack(String),
}

/// One flow: a frozen `type` + wiring, or authored `kinetics` (mirrors `FlowSpec`).
#[derive(Debug, Clone, PartialEq)]
pub struct FlowSpec {
    pub id: String,
    pub type_: Option<String>,
    pub priority: i64,
    pub wiring: Vec<(String, String)>,
    pub kinetics: Option<KineticsSpec>,
    pub params: Option<ParamsSpec>,
}

/// A constant forcing schedule (mirrors `ForcingSpec`; Step-0 constants only).
#[derive(Debug, Clone, PartialEq)]
pub struct ForcingSpec {
    pub const_: NumericField,
}

/// A reusable **domain / species bundle**: stocks + flows + forcings + parameters
/// (mirrors Python `authoring.schema.BundleSpec`, Phase 9, Step 6). A scenario
/// [`ScenarioSpec::includes`] one or more bundle files; [`crate::compose::apply_includes`]
/// merges each bundle's declarations into the scenario's flat graph.
///
/// A bundle carries **no run config** (integrator/dt/steps/name/rng_seed) and **no
/// nested** `includes` — both are rejected by the allowed-key set below (exactly
/// `[parameters, stocks, flows, forcings]`), the `extra="forbid"` analogue: run config
/// lives only in the top-level scenario, and includes are flat, one level deep.
#[derive(Debug, Clone, PartialEq)]
pub struct BundleSpec {
    pub parameters: Vec<(String, f64)>,
    pub stocks: Vec<StockSpec>,
    pub flows: Vec<FlowSpec>,
    pub forcings: Vec<(String, ForcingSpec)>,
}

impl BundleSpec {
    /// Validate and bind a parsed bundle document into a [`BundleSpec`] (the
    /// `BundleSpec.model_validate` analogue). The allowed-key set is exactly
    /// `[parameters, stocks, flows, forcings]`, so a stray `steps:`/`integrator:` (run
    /// config) or a nested `includes:` is a schema error — the two Python
    /// bundle-schema-reject cases.
    pub fn from_yaml(doc: &YamlValue) -> Result<BundleSpec, AuthoringError> {
        let entries = doc.as_mapping("bundle")?;
        reject_unknown_keys(
            entries,
            &["parameters", "stocks", "flows", "forcings"],
            "bundle",
        )?;
        let parameters = match field(entries, "parameters") {
            None => Vec::new(),
            Some(value) => parse_parameters(value)?,
        };
        let stocks = parse_stock_list(entries, "bundle.stocks")?;
        let flows = parse_flow_list(entries, "bundle.flows")?;
        let forcings = match field(entries, "forcings") {
            None => Vec::new(),
            Some(value) => parse_forcings(value)?,
        };
        Ok(BundleSpec {
            parameters,
            stocks,
            flows,
            forcings,
        })
    }
}

/// A whole authored scenario (mirrors `ScenarioSpec`).
#[derive(Debug, Clone, PartialEq)]
pub struct ScenarioSpec {
    pub name: String,
    pub integrator: String,
    pub dt: f64,
    pub steps: u64,
    pub rng_seed: u64,
    /// Bundle-file paths to compose in (Step 6), resolved relative to the scenario
    /// file's directory. Merged by [`crate::compose::apply_includes`] before interpret.
    pub includes: Vec<String>,
    pub parameters: Vec<(String, f64)>,
    pub stocks: Vec<StockSpec>,
    pub flows: Vec<FlowSpec>,
    pub forcings: Vec<(String, ForcingSpec)>,
}

impl ScenarioSpec {
    /// Validate and bind a parsed document into a [`ScenarioSpec`] (the
    /// `ScenarioSpec.model_validate` analogue).
    pub fn from_yaml(doc: &YamlValue) -> Result<ScenarioSpec, AuthoringError> {
        let entries = doc.as_mapping("scenario")?;
        reject_unknown_keys(
            entries,
            &[
                "name",
                "integrator",
                "dt",
                "steps",
                "rng_seed",
                "includes",
                "parameters",
                "stocks",
                "flows",
                "forcings",
            ],
            "scenario",
        )?;
        let name = require_str(entries, "name", "scenario")?;
        let integrator = require_str(entries, "integrator", "scenario")?;
        let dt = require_f64(entries, "dt", "scenario")?;
        let steps = require_u64(entries, "steps", "scenario")?;
        let rng_seed = opt_u64(entries, "rng_seed", "scenario")?.unwrap_or(0);
        let includes = match field(entries, "includes") {
            None => Vec::new(),
            Some(value) => value
                .as_sequence("scenario.includes")?
                .iter()
                .map(|v| scalar_str(v, "scenario.includes item"))
                .collect::<Result<Vec<_>, _>>()?,
        };
        let parameters = match field(entries, "parameters") {
            None => Vec::new(),
            Some(value) => parse_parameters(value)?,
        };
        // `stocks`/`flows` are OPTIONAL (Step 6, carry-forward i): a scenario may omit
        // its inline stocks/flows when it composes them from `includes` (the frozen
        // `crew_station.yaml` has neither). The Python `ScenarioSpec` relaxed these to
        // `Field(default_factory=list)`; this mirror matches.
        let stocks = parse_stock_list(entries, "scenario.stocks")?;
        let flows = parse_flow_list(entries, "scenario.flows")?;
        let forcings = match field(entries, "forcings") {
            None => Vec::new(),
            Some(value) => parse_forcings(value)?,
        };
        Ok(ScenarioSpec {
            name,
            integrator,
            dt,
            steps,
            rng_seed,
            includes,
            parameters,
            stocks,
            flows,
            forcings,
        })
    }
}

/// Parse an optional `stocks:` sequence (shared by `ScenarioSpec` and `BundleSpec`);
/// an absent key is an empty list (Step 6, carry-forward i).
fn parse_stock_list(
    entries: &[(String, YamlValue)],
    ctx: &str,
) -> Result<Vec<StockSpec>, AuthoringError> {
    match field(entries, "stocks") {
        None => Ok(Vec::new()),
        Some(value) => value
            .as_sequence(ctx)?
            .iter()
            .map(parse_stock)
            .collect::<Result<Vec<_>, _>>(),
    }
}

/// Parse an optional `flows:` sequence (shared by `ScenarioSpec` and `BundleSpec`); an
/// absent key is an empty list.
fn parse_flow_list(
    entries: &[(String, YamlValue)],
    ctx: &str,
) -> Result<Vec<FlowSpec>, AuthoringError> {
    match field(entries, "flows") {
        None => Ok(Vec::new()),
        Some(value) => value
            .as_sequence(ctx)?
            .iter()
            .map(parse_flow)
            .collect::<Result<Vec<_>, _>>(),
    }
}

fn parse_parameters(value: &YamlValue) -> Result<Vec<(String, f64)>, AuthoringError> {
    let mut out = Vec::new();
    for (name, v) in value.as_mapping("scenario.parameters")? {
        out.push((name.clone(), scalar_f64(v, &format!("parameter {name:?}"))?));
    }
    Ok(out)
}

fn parse_stock(value: &YamlValue) -> Result<StockSpec, AuthoringError> {
    let entries = value.as_mapping("stock")?;
    reject_unknown_keys(
        entries,
        &[
            "id",
            "domain",
            "quantity",
            "kind",
            "amount",
            "composition",
            "unclamped",
            "extinction_threshold",
        ],
        "stock",
    )?;
    let id = require_str(entries, "id", "stock")?;
    let ctx = format!("stock {id:?}");
    let composition = match field(entries, "composition") {
        None => None,
        Some(v) => {
            let mut comp = Vec::new();
            for (q, coeff) in v.as_mapping(&format!("{ctx} composition"))? {
                comp.push((q.clone(), scalar_f64(coeff, &format!("{ctx} composition {q:?}"))?));
            }
            Some(comp)
        }
    };
    Ok(StockSpec {
        id: id.clone(),
        domain: require_str(entries, "domain", &ctx)?,
        quantity: require_str(entries, "quantity", &ctx)?,
        kind: require_str(entries, "kind", &ctx)?,
        amount: require_numeric(entries, "amount", &ctx)?,
        composition,
        unclamped: opt_bool(entries, "unclamped", &ctx)?.unwrap_or(false),
        extinction_threshold: opt_f64(entries, "extinction_threshold", &ctx)?.unwrap_or(0.0),
    })
}

fn parse_flow(value: &YamlValue) -> Result<FlowSpec, AuthoringError> {
    let entries = value.as_mapping("flow")?;
    reject_unknown_keys(
        entries,
        &["id", "type", "priority", "wiring", "kinetics", "params"],
        "flow",
    )?;
    let id = require_str(entries, "id", "flow")?;
    let ctx = format!("flow {id:?}");
    let type_ = opt_str(entries, "type", &ctx)?;
    let priority = opt_i64(entries, "priority", &ctx)?.unwrap_or(0);
    let wiring = match field(entries, "wiring") {
        None => Vec::new(),
        Some(v) => {
            let mut w = Vec::new();
            for (k, val) in v.as_mapping(&format!("{ctx} wiring"))? {
                w.push((k.clone(), scalar_str(val, &format!("{ctx} wiring {k:?}"))?));
            }
            w
        }
    };
    let kinetics = match field(entries, "kinetics") {
        None => None,
        Some(v) => Some(parse_kinetics(v, &ctx)?),
    };
    let params = match field(entries, "params") {
        None => None,
        Some(v) => Some(parse_params(v, &ctx)?),
    };

    // The `type`-xor-`kinetics` exclusivity + no-wiring-on-kinetics validation
    // (mirrors `FlowSpec._check_type_xor_kinetics`).
    if type_.is_some() == kinetics.is_some() {
        return Err(AuthoringError::new(format!(
            "{ctx}: give exactly one of 'type' (a frozen flow type) or 'kinetics' \
             (an authored rate×stoichiometry flow)"
        )));
    }
    if kinetics.is_some() && !wiring.is_empty() {
        return Err(AuthoringError::new(format!(
            "{ctx}: an authored 'kinetics' flow takes no 'wiring' (its stoichiometry \
             names stocks directly)"
        )));
    }

    Ok(FlowSpec {
        id,
        type_,
        priority,
        wiring,
        kinetics,
        params,
    })
}

fn parse_kinetics(value: &YamlValue, ctx: &str) -> Result<KineticsSpec, AuthoringError> {
    let entries = value.as_mapping(&format!("{ctx} kinetics"))?;
    reject_unknown_keys(entries, &["rate", "stoichiometry"], &format!("{ctx} kinetics"))?;
    let rate = require_str(entries, "rate", &format!("{ctx} kinetics"))?;
    let mut stoichiometry = Vec::new();
    for (stock, coeff) in require_field(entries, "stoichiometry", &format!("{ctx} kinetics"))?
        .as_mapping(&format!("{ctx} kinetics stoichiometry"))?
    {
        stoichiometry.push((
            stock.clone(),
            scalar_f64(coeff, &format!("{ctx} stoichiometry {stock:?}"))?,
        ));
    }
    Ok(KineticsSpec { rate, stoichiometry })
}

fn parse_params(value: &YamlValue, ctx: &str) -> Result<ParamsSpec, AuthoringError> {
    match value {
        YamlValue::Scalar { text, .. } => Ok(ParamsSpec::Named(text.clone())),
        YamlValue::Mapping(_) => {
            let entries = value.as_mapping(&format!("{ctx} params"))?;
            reject_unknown_keys(entries, &["pack"], &format!("{ctx} params"))?;
            let pack = require_str(entries, "pack", &format!("{ctx} params"))?;
            Ok(ParamsSpec::Pack(pack))
        }
        _ => Err(AuthoringError::new(format!(
            "{ctx}: 'params' must be a param-set name or a {{pack: …}} reference"
        ))),
    }
}

fn parse_forcings(value: &YamlValue) -> Result<Vec<(String, ForcingSpec)>, AuthoringError> {
    let mut out = Vec::new();
    for (name, v) in value.as_mapping("scenario.forcings")? {
        let entries = v.as_mapping(&format!("forcing {name:?}"))?;
        reject_unknown_keys(entries, &["const"], &format!("forcing {name:?}"))?;
        let const_ = require_numeric(entries, "const", &format!("forcing {name:?}"))?;
        out.push((name.clone(), ForcingSpec { const_ }));
    }
    Ok(out)
}

// --------------------------------------------------------------------------- #
// Field accessors + `extra="forbid"` enforcement.                              #
// --------------------------------------------------------------------------- #

fn field<'a>(entries: &'a [(String, YamlValue)], key: &str) -> Option<&'a YamlValue> {
    entries.iter().find(|(k, _)| k == key).map(|(_, v)| v)
}

fn require_field<'a>(
    entries: &'a [(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<&'a YamlValue, AuthoringError> {
    field(entries, key).ok_or_else(|| AuthoringError::new(format!("{ctx}: missing key {key:?}")))
}

fn reject_unknown_keys(
    entries: &[(String, YamlValue)],
    allowed: &[&str],
    ctx: &str,
) -> Result<(), AuthoringError> {
    for (key, _) in entries {
        if !allowed.contains(&key.as_str()) {
            return Err(AuthoringError::new(format!(
                "{ctx}: unknown key {key:?} (allowed: {allowed:?})"
            )));
        }
    }
    Ok(())
}

fn scalar_str(value: &YamlValue, ctx: &str) -> Result<String, AuthoringError> {
    let (text, _) = value.as_scalar(ctx)?;
    Ok(text.to_string())
}

fn scalar_f64(value: &YamlValue, ctx: &str) -> Result<f64, AuthoringError> {
    let (text, _) = value.as_scalar(ctx)?;
    text.parse::<f64>()
        .map_err(|_| AuthoringError::new(format!("{ctx}: {text:?} is not a number")))
}

fn require_str(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<String, AuthoringError> {
    scalar_str(require_field(entries, key, ctx)?, &format!("{ctx}.{key}"))
}

fn opt_str(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<Option<String>, AuthoringError> {
    match field(entries, key) {
        None => Ok(None),
        Some(v) => Ok(Some(scalar_str(v, &format!("{ctx}.{key}"))?)),
    }
}

fn require_f64(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<f64, AuthoringError> {
    scalar_f64(require_field(entries, key, ctx)?, &format!("{ctx}.{key}"))
}

fn opt_f64(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<Option<f64>, AuthoringError> {
    match field(entries, key) {
        None => Ok(None),
        Some(v) => Ok(Some(scalar_f64(v, &format!("{ctx}.{key}"))?)),
    }
}

fn require_u64(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<u64, AuthoringError> {
    let ctx2 = format!("{ctx}.{key}");
    let (text, _) = require_field(entries, key, ctx)?.as_scalar(&ctx2)?;
    text.parse::<u64>()
        .map_err(|_| AuthoringError::new(format!("{ctx2}: {text:?} is not a non-negative integer")))
}

fn opt_u64(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<Option<u64>, AuthoringError> {
    match field(entries, key) {
        None => Ok(None),
        Some(_) => Ok(Some(require_u64(entries, key, ctx)?)),
    }
}

fn opt_i64(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<Option<i64>, AuthoringError> {
    match field(entries, key) {
        None => Ok(None),
        Some(v) => {
            let ctx2 = format!("{ctx}.{key}");
            let (text, _) = v.as_scalar(&ctx2)?;
            Ok(Some(text.parse::<i64>().map_err(|_| {
                AuthoringError::new(format!("{ctx2}: {text:?} is not an integer"))
            })?))
        }
    }
}

fn opt_bool(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<Option<bool>, AuthoringError> {
    match field(entries, key) {
        None => Ok(None),
        Some(v) => {
            let ctx2 = format!("{ctx}.{key}");
            let (text, _) = v.as_scalar(&ctx2)?;
            match text {
                "true" => Ok(Some(true)),
                "false" => Ok(Some(false)),
                other => Err(AuthoringError::new(format!(
                    "{ctx2}: {other:?} is not a boolean (use true/false)"
                ))),
            }
        }
    }
}

/// Bind a `number | expression` union field (the `float | str` union): a bare YAML-1.1
/// number → [`NumericField::Number`]; a quoted or non-number scalar →
/// [`NumericField::Expr`] (the pyyaml float-vs-str split).
fn require_numeric(
    entries: &[(String, YamlValue)],
    key: &str,
    ctx: &str,
) -> Result<NumericField, AuthoringError> {
    let ctx2 = format!("{ctx}.{key}");
    let (text, quoted) = require_field(entries, key, ctx)?.as_scalar(&ctx2)?;
    if !quoted && is_yaml_number(text) {
        Ok(NumericField::Number(text.parse::<f64>().map_err(|_| {
            AuthoringError::new(format!("{ctx2}: {text:?} is not a number"))
        })?))
    } else {
        Ok(NumericField::Expr(text.to_string()))
    }
}
